### PR TITLE
Ability to choose TLS version in LDAP transactions

### PR DESCRIPTION
**Pull request by maturbet-volta**
For security reasons, some firewalls filter out older versions of TLS, even if the server and client accept it.

Add the ability to choose which version of TLS to use

### CODE REVIEW

File: src/AuthLDAP.php [hunk 1]
```diff
@@ -2950,6 +2956,7 @@ public static function connectToServer(
         $deref_options = 0,
         $tls_certfile = "",
         $tls_keyfile = "",
+        $tls_version = "",
```
Comment (cedric-anne) [inline src/AuthLDAP.php:2959]: This new parameter should be placed after existing ones, to ensure that existing calls from plugins will still work as expected.

Comment (cedric-anne) [inline src/AuthLDAP.php:2959]: This new parameter should be placed after existing ones, to ensure that existing calls from plugins will still work as expected.

File: install/mysql/glpi-10.1.0-empty.sql [hunk 2]
```diff
@@ -117,6 +117,7 @@ CREATE TABLE `glpi_authldaps` (
   `inventory_domain` varchar(255) DEFAULT NULL,
   `tls_certfile` text,
   `tls_keyfile` text,
+  `tls_version` tinyint DEFAULT NULL,
```
Comment (trasher) [inline install/mysql/glpi-10.1.0-empty.sql:120]: You have to add the new field in migrations

**Review by trasher — APPROVED** [review]

**Review by orthagh — APPROVED** [review]

File: install/mysql/glpi-10.1.0-empty.sql [hunk 3]
```diff
@@ -119,6 +119,7 @@ CREATE TABLE `glpi_authldaps` (
   `tls_keyfile` text,
   `use_bind` tinyint NOT NULL DEFAULT '1',
   `timeout` int NOT NULL DEFAULT '10',
+  `tls_version` tinyint DEFAULT NULL,
```
Comment (cedric-anne) [inline install/mysql/glpi-10.1.0-empty.sql:122]:
```suggestion
  `tls_version` varchar(10) DEFAULT NULL,
```

File: install/migrations/update_10.0.x_to_10.1.0/tls_version.php [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ * ---------------------------------------------------------------------
+ */
+
+if (!$DB->fieldExists('glpi_authldaps', 'tls_version', false)) {
+    $migration->addField('glpi_authldaps', 'tls_version', 'tinyint DEFAULT NULL', ['after' => 'timeout']);
```
Comment (cedric-anne) [inline install/migrations/update_10.0.x_to_10.1.0/tls_version.php:37]:
```suggestion
    $migration->addField('glpi_authldaps', 'tls_version', 'varchar(10) DEFAULT NULL', ['after' => 'timeout']);
```

File: src/AuthLDAP.php [hunk 5]
```diff
@@ -624,9 +624,12 @@ public function showFormAdvancedConfig()
         echo "</td>";
         echo "</tr>";
 
-        echo "<tr class='tab_bg_1'><td><label for='timeout'>" . __('Timeout') . "</label></td>";
-        echo "<td colspan='3'>";
-
+        echo "<tr class='tab_bg_1'>";
+        echo "<td>" . __('TLS Version') . "</td><td>";
+        Dropdown::showFromArray('tls_version', ['1.0', '1.1', '1.2', '1.3'], ['value' => $this->fields["tls_version"]]);
```
Comment (cedric-anne) [inline src/AuthLDAP.php:629]:
```suggestion
        Dropdown::showFromArray('tls_version', ['1.0' => '1.0', '1.1' => '1.1', '1.2' => '1.2', '1.3' => '1.3'], ['value' => $this->fields["tls_version"]]);
```

File: src/AuthLDAP.php [hunk 6]
```diff
@@ -2969,6 +2976,10 @@ public static function connectToServer(
                 @ldap_set_option(null, LDAP_OPT_X_TLS_KEYFILE, $tls_keyfile);
             }
 
+            if (!empty($tls_version)) {
+                @ldap_set_option(null, LDAP_OPT_X_TLS_CIPHER_SUITE, "NORMAL:!VERS-TLS1." . $tls_version);
```
Comment (cedric-anne) [inline src/AuthLDAP.php:2980]:
```suggestion
                @ldap_set_option(null, LDAP_OPT_X_TLS_CIPHER_SUITE, "NORMAL:!VERS-TLS" . $tls_version);
```

**Review by cedric-anne — CHANGES_REQUESTED** [review]
Proposed logic works only with versions of TLS that follows the `1.x` pattern where `x` is a unique digit. I guess it may be possible to have `2.x` versions in the future, or maybe `1.x.y` versions, so I would prefer to explicitely store the version string in DB.

File: src/AuthLDAP.php (hunk 5 above)
Comment (cedric-anne) [inline src/AuthLDAP.php:629]: Maybe a `'' => Dropdown::EMPTY_VALUE` should also be added on top of the proposed versions.

Comment (maturbet-volta) [reply src/AuthLDAP.php]: For more readability, I have used a constant.

**Comment by maturbet-volta** [comment]
I just corrected the inversion between excluding and adding the TLS version.

**Review by cedric-anne — COMMENTED** [review]
Seems OK to me, but could you please add a link to some documentation that explains expected values for this option ?

**Comment by maturbet-volta** [comment]
The GNU documentation for using TLS : https://gnutls.org/manual/gnutls.html#Using-Priority-Strings

**Comment by maturbet-volta** [comment]
Functionality also useful to bypass the libldap issue ([libldap support for TLS SNI](https://bugs.openldap.org/show_bug.cgi?id=9176))

**Review by cedric-anne — APPROVED** [review]

**Pull request merged**